hollow/src/main/java/com/netflix/hollow/api/producer/metrics/AbstractProducerMetricsListener.java

```diff
@@ -0,0 +1,125 @@
+
+
+    public AbstractProducerMetricsListener() {
+        consecutiveFailures = 0l;
+        lastCycleSuccessTimeNanoOptional = OptionalLong.empty();
+        lastAnnouncementSuccessTimeNanoOptional = OptionalLong.empty();
+    }
+
+    @Override
+    public void onCycleStart(long version) {
+        cycleMetricsBuilder = new CycleMetrics.Builder();
+    }
+
+    @Override
+    public void onAnnouncementStart(long version) {
+        announcementMetricsBuilder = new AnnouncementMetrics.Builder();
+    }
+
+    /**
+     * Reports metrics for when cycle is skipped due to reasons such as the producer not being the leader in a multiple-producer setting.
+     * In a multiple producer setting, leader election typically favors long-lived leaders to avoid producer runs from frequently requiring
+     * to reload the full state before publishing data. When a cycle is skipped because the producer wasn't primary, the current value of
+     * no. of consecutive failures and most recent cycle success time are retained, and no cycle status (success or fail) is reported.
+     * @param reason Reason why the run was skipped
+     */
+    @Override
+    public void onCycleSkip(CycleSkipReason reason) {
+        cycleMetricsBuilder.setConsecutiveFailures(consecutiveFailures);
+
+        if (lastCycleSuccessTimeNanoOptional.isPresent())
+            cycleMetricsBuilder.setLastCycleSuccessTimeNanoOptional((lastCycleSuccessTimeNanoOptional.getAsLong()));
+
+        // isCycleSuccess and cycleDurationMillis are not set for skipped cycles
+        cycleMetricsBuilder.setConsecutiveFailures(consecutiveFailures);
+
+        cycleMetricsReporting(cycleMetricsBuilder.build());
+    }
+
+    /**
+     * Reports announcement-related metrics.
+     * @param status Indicates whether the announcement succeeded of failed
+     * @param readState Hollow data state that is being published, used in this method for computing data size
+     * @param version Version of data that was announced
+     * @param elapsed Announcement start to end duration
+     */
+    @Override
+    public void onAnnouncementComplete(com.netflix.hollow.api.producer.Status status, HollowProducer.ReadState readState, long version, Duration elapsed) {
+        boolean isAnnouncementSuccess = false;
+        long dataSizeBytes = 0l;
+
+        if (status.getType() == com.netflix.hollow.api.producer.Status.StatusType.SUCCESS) {
+            isAnnouncementSuccess = true;
+            lastAnnouncementSuccessTimeNanoOptional = OptionalLong.of(System.nanoTime());
+        }
+
+        HollowReadStateEngine stateEngine = readState.getStateEngine();
+        dataSizeBytes = stateEngine.calcApproxDataSize();
+
+        announcementMetricsBuilder
+                .setDataSizeBytes(dataSizeBytes)
+                .setIsAnnouncementSuccess(isAnnouncementSuccess)
+                .setAnnouncementDurationMillis(elapsed.toMillis());
+
+        if (lastAnnouncementSuccessTimeNanoOptional.isPresent())
```

Comment:
One way to avoid the `isPresent` and `get` dance is to use `ifPresent`. I suspect you can pass in a method reference in this case. Up to you.

Reply:
Neat, done